transports/tcp/src/lib.rs

```diff
@@ -748,6 +682,15 @@ where
             }
         }
 
+        if let Some(event) = me.queued_events.pop_front() {
+            return Poll::Ready(Some(event));
+        }
+
+        if me.is_closed {
+            // Terminate the stream if the listener closed and all remaining events have been reported.
+            return Poll::Ready(None);
+        }
```

Comment:
We now have two return paths here that don't register their own waker. You'll need to store a waker in the listener before you return `Poll::Pending` at the bottom.

Reply:
I lack expertise on this particular issue. Instead I have aligned the impl for the `tcp` transport with the impl for the [`quic` transport](https://github.com/libp2p/rust-libp2p/blob/4b41f5a9942add3f7a815bd2b79cb82eaed1ee05/transports/quic/src/transport.rs#L316), they are pretty similar.

Reply:
@elenaf9 Did you consider this case when implementing QUIC? I think we may have a potential unwoken task here:

- Assume 1 active listener
- Assume the listener had no work and returned `Pending`
- We call `close` on the listener
- Without an explicit waker, I believe the close will only happen once the listener gets polled again due to another event

Reply:
Well, it appears this PR can't be merged before this question is cleared out.

Reply:
We can move forward here by just adding the waker. Worst case, we have an unnecessary wake-up.

- Make a new field inside the listener, holding an `Option<Waker>`
- Call it `close_listener_waker`
- Set it before you return `Poll::Pending` at the bottom
- Use `take` to remove the listener within `close` and if it is present, call `wake` on it

Reply:
@jxs @mxinden perhaps you want to take a look at this too.

Reply:
In my eyes, the suggestion above by @thomaseizinger is the way to go.

Reply:
Good catch @thomaseizinger. Yes you are right, we need the explicit waker here.